messaging/schema.py

```diff
@@ -0,0 +1,243 @@
+            raise MessageAlreadySentError(
+                "Cannot update because the message has already been sent."
+            )
+
+        if "project_id" in data:
+            data["project_id"] = get_obj_if_user_can_administer(
+                info, data.pop("project_id"), Project
+            )
+
+        if "event_id" in data:
+            data["event"] = get_obj_if_user_can_administer(
+                info, data.pop("event_id"), Event
+            )
+        if "occurrence_ids" in data:
+            occurrences = [
+                get_obj_if_user_can_administer(info, occurrence_id, Occurrence)
+                for occurrence_id in data.pop("occurrence_ids", [])
+            ]
+        else:
+            occurrences = None
+
+        validate_event_and_occurrences(
+            data["event"] if "event" in data else message.event,
+            occurrences if occurrences is not None else message.occurrences.all(),
+        )
+
+        update_object_with_translations(message, data)
+
+        if occurrences is not None:
+            message.occurrences.set(occurrences)
+
+        logger.info(
+            f"user {info.context.user.uuid} updated message {message} with data {kwargs}"  # noqa: E501
+        )
+
+        return UpdateMessageMutation(message=message)
+
+
+def validate_event_and_occurrences(event, occurrences):
+    if not occurrences:
+        return
+
+    if any(occurrence.event != event for occurrence in occurrences):
+        raise DataValidationError(
+            "All of the occurrences do not belong to the same event."
+        )
+
+
+class SendMessageMutation(graphene.relay.ClientIDMutation):
+    class Input:
+        id = graphene.GlobalID()
+
+    message = graphene.Field(MessageNode)
+
+    @classmethod
+    @project_user_required
+    def mutate_and_get_payload(cls, root, info, **kwargs):
+        message = get_obj_if_user_can_administer(info, kwargs["id"], Message)
+        if message.sent_at:
+            raise MessageAlreadySentError(
+                "Cannot send because the message has already been sent."
+            )
+
+        message.send()
+
+        logger.info(f"user {info.context.user.uuid} sent message {message}")
+
+        return SendMessageMutation(message=message)
+
+
+class DeleteMessageMutation(graphene.relay.ClientIDMutation):
+    class Input:
+        id = graphene.GlobalID()
+
+    @classmethod
+    @project_user_required
+    def mutate_and_get_payload(cls, root, info, **kwargs):
+        message = get_obj_if_user_can_administer(info, kwargs["id"], Message)
+        if message.sent_at:
+            raise MessageAlreadySentError(
+                "Cannot delete because the message has already been sent."
+            )
+
+        log_text = f"user {info.context.user.uuid} deleted message {message}"
+        message.delete()
+
+        logger.info(log_text)
+
+        return DeleteMessageMutation()
+
+
+class Query:
+    message = relay.Node.Field(MessageNode)
```

Comment:
Are these queries admin only?

Reply:
Indeed they should be (but are not), will fix.